plugins/inputs/firehose/firehose.go

```diff
@@ -0,0 +1,178 @@
+	"net"
+	"net/http"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/internal"
+	tlsint "github.com/influxdata/telegraf/plugins/common/tls"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+// Firehose is an input plugin that collects external metrics sent via HTTP from AWS Data Firhose
+type Firehose struct {
+	ServiceAddress string          `toml:"service_address"`
+	Paths          []string        `toml:"paths"`
+	ReadTimeout    config.Duration `toml:"read_timeout"`
+	WriteTimeout   config.Duration `toml:"write_timeout"`
+	AccessKey      config.Secret   `toml:"access_key"`
+	ParameterTags  []string        `toml:"parameter_tags"`
+	Log            telegraf.Logger `toml:"-"`
+
+	tlsint.ServerConfig
+	tlsConf *tls.Config
+
+	once sync.Once
+
+	listener net.Listener
+	server   http.Server
+
+	parser telegraf.Parser
+	acc    telegraf.Accumulator
+}
+
+func (*Firehose) SampleConfig() string {
+	return sampleConfig
+}
+
+func (*Firehose) Gather(telegraf.Accumulator) error {
+	return nil
+}
+
+func (f *Firehose) SetParser(parser telegraf.Parser) {
+	f.parser = parser
+}
+
+func (f *Firehose) Init() error {
+	if f.ServiceAddress == "" {
+		f.ServiceAddress = ":8080"
+	}
+	if len(f.Paths) == 0 {
+		f.Paths = []string{"/telegraf"}
+	}
+
+	var err error
+	f.tlsConf, err = f.ServerConfig.TLSConfig()
+	return err
+}
+
+// Start starts the http listener service.
+func (f *Firehose) Start(acc telegraf.Accumulator) error {
+	f.acc = acc
+
+	var err error
+	if f.tlsConf != nil {
+		f.listener, err = tls.Listen("tcp", f.ServiceAddress, f.tlsConf)
+	} else {
+		f.listener, err = net.Listen("tcp", f.ServiceAddress)
+	}
+	if err != nil {
+		return err
+	}
+
+	f.server = http.Server{
+		Addr:         f.ServiceAddress,
+		Handler:      f,
+		ReadTimeout:  time.Duration(f.ReadTimeout),
+		WriteTimeout: time.Duration(f.WriteTimeout),
+		TLSConfig:    f.tlsConf,
+	}
+
+	go func() {
+		if err := f.server.Serve(f.listener); err != nil {
+			if !errors.Is(err, net.ErrClosed) {
+				f.Log.Errorf("Starting server failed: %v", err)
+			}
+		}
+	}()
+
+	f.Log.Infof("Listening on %s", f.listener.Addr().String())
+
+	return nil
+}
+
+// Stop cleans up all resources
+func (f *Firehose) Stop() {
+	err := f.server.Shutdown(context.Background())
+	if err != nil {
+		f.Log.Errorf("Shutting down server failed: %v", err)
+	}
+}
+
+func (f *Firehose) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	if !slices.Contains(f.Paths, req.URL.Path) {
+		res.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	r := f.handleRequest(req)
+
+	if err := r.sendResponse(res); err != nil {
+		f.Log.Errorf("Sending response failed: %v", err)
+	}
+}
+
+func (f *Firehose) handleRequest(req *http.Request) (r *request) {
+	var err error
+	if r, err = newRequest(req); err != nil {
+		f.Log.Errorf("Creating request object failed: %v", err)
+		return r
+	}
```

Comment:
How about returning the actual error here and do the logging in `ServeHTTP`? This will save a few lines I think
```suggestion
func (f *Firehose) handleRequest(req *http.Request) (*request, error) {
	r, err := newRequest(req)
	if err != nil {
		return fmt.Errorf("creating request failed: %w", err)
	}
```
of course you need to modify the code below and in `ServeHTTP`...

Reply:
I tried implementing this but doesn’t help. The line count remains the same as we need to introduce err checking in the `ServeHTTP` function. I made some other changes which saves us a couple lines though